tests/test_BT_piecewise.py

```diff
@@ -0,0 +1,560 @@
+        (paramX_per_toa.value - test_val[0].value), 0, atol=1e-6, rtol=0
+    )
+    do_toas_reference_second_piecewise_parameter = np.isclose(
+        (paramX_per_toa.value - test_val[1].value), 0, atol=1e-6, rtol=0
+    )
+    do_toas_reference_piecewise_parameter = [
+        do_toas_reference_first_piecewise_parameter,
+        do_toas_reference_second_piecewise_parameter,
+    ]
+    do_arrays_match = np.array_equal(
+        do_toas_reference_piecewise_parameter, should_toa_reference_piecewise_parameter
+    )
+    assert do_arrays_match
+
+
+def test_problematic_group_indexes_and_ranges(build_model):
+    # Test to flag issues with problematic group indexes
+    m_piecewise = build_model
+    with pytest.raises(Exception):
+        assert m_piecewise.add_group_range(
+            m_piecewise.START.value, m_piecewise.FINISH.value, j=-1
+        )
+        assert m_piecewise.add_group_range(
+            m_piecewise.FINISH.value, m_piecewise.START.value, j=1
+        )
+        assert m_piecewise.add_group_range(
+            m_piecewise.START.value, m_piecewise.FINISH.value, j=10000
+        )
+
+
+def add_groups_and_make_toas(build_model, build_piecewise_model_with_two_pieces, param):
+    # function to build the models for specific edge cases
+    spacing = (build_model.FINISH.value - build_model.START.value) / 2
+    if param == "non-overlapping complete group coverage":
+        model = build_piecewise_model_with_two_pieces
+        model.setup()
+        # print(model.as_parfile())
+        toas = make_generic_toas(model)
+        return model, toas
+    elif param == "overlapping groups":
+        model2 = build_model
+        model2.remove_range(0)
+        model2.add_group_range(
+            model2.START.value - 1, model2.START.value + spacing + 100, j=0
+        )
+        model2.add_group_range(
+            model2.START.value + spacing - 100, model2.FINISH.value + 1, j=1
+        )
+        model2.add_piecewise_param("T0", "d", model2.T0.value + 1e-5, 0)
+        model2.add_piecewise_param("T0", "d", model2.T0.value + 2e-5, 1)
+        model2.add_piecewise_param("A1", "ls", model2.A1.value + 1e-5, 0)
+        model2.add_piecewise_param("A1", "ls", model2.A1.value + 2e-5, 1)
+        model2.setup()
+        toas = make_generic_toas(model2)
+        return model2, toas
+    elif param == "non-complete group coverage":
+        model3 = build_model
+        model3.remove_range(0)
+        model3.add_group_range(
+            model3.START.value + spacing, model3.FINISH.value + 1, j=0
+        )
+        model3.add_piecewise_param("T0", "d", model3.T0.value + 1e-5, 0)
+        model3.setup()
+        toas = make_generic_toas(model3)
+        return model3, toas
+
+
+def make_generic_toas(model):
+    # makes toas to go with the edge cases
+    return pint.simulation.make_fake_toas_uniform(
+        model.START.value, model.FINISH.value - 1, 20, model
+    )
+
+
+def convert_int_into_index(i):
```

Comment:
Isn't there a utility function to do this in the `BTpiecewise` code?

Reply:
It's removed now, all it was doing is making 1->0001 so I just declare them where necessary now rather than using a function